server/hotrod/src/main/java/org/infinispan/server/hotrod/HotRodServer.java

```diff
@@ -673,26 +674,33 @@ public void update(boolean enabled, boolean indexing, boolean syncListener) {
       }
 
       @TopologyChanged
-      public void topologyChanged(TopologyChangedEvent<Address, ServerAddress> event) {
-         boolean success = false;
-         while (!success && addressCache.getStatus().allowInvocations()) {
-            try {
-               // No need for a timeout here, the cluster executor has a default timeout
-               CompletableFuture<Void> future = clusterExecutor.submitConsumer(new CheckAddressTask(addressCache.getName(),
-                     clusterAddress), (a, v, t) -> {
-                  if (t != null) {
-                     throw new CacheException(t);
-                  }
-                  if (!v) {
-                     log.debugf("Re-adding %s to the topology cache", clusterAddress);
-                     addressCache.putAsync(clusterAddress, address);
-                  }
-               });
-               future.get();
-               success = true;
-            } catch (Throwable e) {
-               log.debug("Error re-adding address to topology cache, retrying", e);
-            }
+      public CompletionStage<Void> topologyChanged(TopologyChangedEvent<Address, ServerAddress> event) {
```

Comment:
Does it make sense to return a completion stage with `@Listener(sync = false)`?

Either way, adding the server address to the topology cache should not block the topology update, so we shouldn't return a completion stage.

Reply:
Oops, didn't notice that. I can remove it and just revert to how it was before. Good catch.

Reply:
And tbh this is harmless currently as the return value is never checked when sync is disabled. - harmless assuming no bugs :)

Reply:
It complicates the code IMO, `recursionTopologyChanged` wouldn't be needed without the future. And since you're doing everything asynchronously I guess you can also remove both the `CompletionStage` return value *and* `sync=false`.

Reply:
Trust me I said I want to change it, it will come in the other JIRA. I was just saying there is no hurry to fix it as it causes no issues.